src/EntityRegistrantDetection/EntityRegistrantDetectionManager.php

```diff
@@ -0,0 +1,305 @@
+						'source'       => $source,
+					];
+					break;
+				case 'block':
+					if ( ! class_exists( 'WP_Block_Type_Registry' ) ) {
+						break;
+					}
+
+					$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $entity );
+
+					if ( empty( $block_type ) || ! $block_type instanceof WP_Block_Type ) {
+						break;
+					}
+
+					$this->blocks_source[ $entity ] = [
+						'name'        => $block_type->name,
+						'title'       => isset( $block_type->title ) ? $block_type->title : '',
+						'description' => isset( $block_type->description ) ? $block_type->description : '',
+						'category'    => isset( $block_type->category ) ? $block_type->category : '',
+						'attributes'  => $block_type->get_attributes(),
+						'is_dynamic'  => $block_type->is_dynamic(),
+						'source'      => $source,
+					];
+					break;
+				case 'shortcode':
+					global $shortcode_tags;
+
+					if ( ! isset( $shortcode_tags[ $entity ] ) ) {
+						break;
+					}
+
+					$this->shortcodes_source[ $entity ] = [
+						'tag'    => $entity,
+						'source' => $source,
+					];
+					break;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get list of registered entities.
+	 *
+	 * @return array List of registered entities.
+	 */
+	public function get_registered_entities() {
+
+		return [
+			'post_types' => $this->post_types_source,
+			'taxonomies' => $this->taxonomies_source,
+			'blocks'     => $this->blocks_source,
+			'shortcodes' => $this->shortcodes_source,
+		];
+	}
+
+	/**
+	 * Adds hooks.
+	 */
+	public function register() {
+
+		$int_min = defined( 'PHP_INT_MIN' ) ? PHP_INT_MIN : ~PHP_INT_MAX; // phpcs:ignore PHPCompatibility.Constants.NewConstants.php_int_minFound
+
+		add_action( 'all', [ $this, 'wrap_hook_callbacks' ], $int_min );
+	}
+
+	/**
+	 * Wrap action/filter callback to given hook to check if callback registering any entity or not.
+	 *
+	 * @global WP_Hook[] $wp_filter
+	 *
+	 * @param string $hook Hook name for action or filter.
+	 *
+	 * @return void
+	 */
+	public function wrap_hook_callbacks( $hook ) {
+
+		global $wp_filter;
+
+		if ( ! isset( $wp_filter[ $hook ] ) || 'shutdown' === $hook ) {
+			return;
+		}
+
+		foreach ( $wp_filter[ $hook ]->callbacks as $priority => &$callbacks ) {
+			foreach ( $callbacks as &$callback ) {
+				if ( ! $callback['function'] instanceof CallbackWrapper ) {
+					$callback['function'] = $this->wrapped_callback(
+						array_merge(
+							$callback,
+							compact( 'priority', 'hook' )
+						)
+					);
+				}
+			}
+		}
+
+	}
+
+	/**
+	 * Wrap the given callable function.
+	 *
+	 * @param array $callback [
+	 *     The callback data.
+	 *     @type callable $function
+	 *     @type int      $accepted_args
+	 *     @type int      $priority
+	 *     @type int      $hook
+	 * ]
+	 *
+	 * @return CallbackWrapper Instance of CallbackWrapper.
+	 */
+	protected function wrapped_callback( $callback ) {
+
+		return $this->injector->make( CallbackWrapper::class, compact( 'callback' ) );
```

Comment:
Injecting the injector should always be avoided whenever possible. There are two viable options for doing so here:

**If `CallbackWrapper` is considered a service or similar:**

Instead of injecting the injector (and hence, giving control to the object of the entire rest of the application), it would be preferable to inject a `CallbackWrapperFactory` instead which is specialized on directly instantiating instances of type `CallbackWrapper`.

Using a factory is always a valid way of decoupling a `new` call, as the injector can override the factory implementation if needed.

**If `CallbackWrapper` is considered a Value Object:**

If we consider `CallbackWrapper` to be a [value object](https://en.wikipedia.org/wiki/Value_object) conceptually (which is a bit far-fetched but not entirely wrong), then it would mean we can freely `new` that value object directly, instead of having it be instantiated via the injector.